### PR TITLE
e2e: avoid non-determinism in app hash check

### PR DESCRIPTION
**Pull request by tychoish**
My best theory is that because the e2e test assertions run while the network continues to make progress, we're occasionally getting apphash errors when new blocks are produced during the test.

This should make sure we're (almost always comparing apples to apples)

### CODE REVIEW

File: test/e2e/tests/app_test.go [hunk 1]
```diff
@@ -44,15 +44,15 @@ func TestApp_Hash(t *testing.T) {
...
-		block, err := client.Block(ctx, nil)
-		require.NoError(t, err)
-		require.EqualValues(t, info.Response.LastBlockAppHash, block.Block.AppHash.Bytes(),
-			"app hash does not match last block's app hash")
-
 		status, err := client.Status(ctx)
 		require.NoError(t, err)
 		require.EqualValues(t, info.Response.LastBlockAppHash, status.SyncInfo.LatestAppHash,
 			"app hash does not match node status")
+
```
Comment (creachadair) [inline test/e2e/tests/app_test.go:51]: I'm trying to understand how changing the order could affect the outcome here. The Status call doesn't seem to have any access to the data we're comparing. Is the idea that calling Status could change what block we get back from the wrong value to the right one?

What am I missing? 🙂

File: test/e2e/tests/app_test.go [hunk 2]
```diff
@@ -44,15 +44,15 @@ func TestApp_Hash(t *testing.T) {
...
+		block, err := client.Block(ctx, &info.Response.LastBlockHeight)
+		require.NoError(t, err)
+		require.EqualValues(t, info.Response.LastBlockAppHash, block.Block.AppHash.Bytes(),
```
Comment (williambanfield) [inline test/e2e/tests/app_test.go:54]: Can we check here that the Block's height matches the `info.Response.LastBlockHeight`?

Comment (cmwaters) [reply test/e2e/tests/app_test.go:54]: We could, however, ensuring that heights match should be something covered by unit tests

File: test/e2e/tests/app_test.go (hunk 1 above)
Comment (cmwaters) [reply test/e2e/tests/app_test.go:51]: I think this is just an issue of atomicity. Beforehand we just asked for the latest block which could have changed (i.e. a new block may have been produced) in between calling `ABCIInfo` and `Block`. This is solved by specifying the height (being the same as was returned in `ABCIInfo` in the call to retrieve the `Block`. However, the problem still persists. It's conceivable that a new block is committed in between `ABCIInfo` and `Status` in which case we will get different hashes. The best solution I think is that we check the heights match before checking the app hashes.
```
status, err := client.Status(ctx)
require.NoError(t, err)
if info.Response.LastBlockHeight == status.SyncInfo.LastBlockHeight {
    require.EqualValues(t, info.Response.LastBlockAppHash, status.SyncInfo.LatestAppHash,
        "app hash does not match node status")
}
```
There are a few atomicity problems in Tendermint. One example is that we don't batch persisting of blocks and state together. This means occasionally the block store is ahead of the state store and we need to have logic on restarts to handle this. Looking forward to the db work fixing some of this

**Review by cmwaters — APPROVED** [review]
LGTM 👍  - I just have the one comment about checking that the height matches before checking the app hashes are the same

**Comment by tychoish** [comment]
Thanks @cmwaters for fielding questions about this! I couldn't have said it better. I also made the change that you suggested.

**Pull request merged**